src/v/redpanda/application.cc

```diff
@@ -2622,6 +2663,18 @@ void application::wire_up_and_start(::stop_signal& app_signal, bool test_mode) {
           "Running with already-established node ID {}",
           config::node().node_id());
         node_id = config::node().node_id().value();
+    } else if (auto id = _node_overrides.node_id(); id.has_value()) {
```

Comment:
I think it makes sense to try to register with the cluster (next conditional branch) in this case as well. For two reasons:
* it will get us fresh features and cluster config snapshots from the controller leader
* it will allow us to reject erroneous configurations (e.g. if the UUID is already registered with a different id).

Not sure if all the code for this is there, but looks like at least the join RPC supports passing existing node ids.

Reply:
Makes sense. I'll refactor the conditionals a little bit.

Reply:
Won't we hit the problem of needing a controller leader if we try to register here? Or do we also want to change `members_manager::handle_join_request` to not require controller leadership when the node is trying to register with a known (node uuid, node id) pair?

https://github.com/redpanda-data/redpanda/blob/74223be8ce5ee37c165bcfcb0187018cdb6653ed/src/v/cluster/members_manager.cc#L1240-L1298

Reply:
Hmm good point, maybe it won't work then.

Reply:
yeah, the exact requirement we're trying to circumvent. i was wondering whether the cluster layer might short circuit somewhere (or be made to short circuit) on previously known <ID,UUID>, but I suppose all the request routing is controller leadership based 🤷